feat(awful: widgets): add 'watch' widget

actionless opened the pull request:
follow-up on https://github.com/awesomeWM/awesome/pull/357

psychon commented on lib/awful/widget/watch.lua:
```diff
@@ -0,0 +1,79 @@
+    stderr=parse_stderr,
+    exit=exit_callback,
+    output_done=output_done_callback
+  })
```
Since my comments elsewhere got lost:

You do not need an `exit_callback`, just `output_done` is enough for this code (that simplifies the code a bit).

actionless replied:
i was using that `easy_async` helper for some time and noticed what some programs require that

psychon replied:
Uhm, what? Any example that I could investigate?

actionless replied:
i hope now it should make more sense

actionless replied:
btw, mb we can move it to `awful.spawn`?

psychon replied:
If you do a PR for that, that might be doable. :-P

actionless replied:
https://github.com/awesomeWM/awesome/pull/668

psychon commented on lib/awful/widget/watch.lua:
```diff
@@ -0,0 +1,79 @@
+
+--- Create a textbox that shows the output of a command and updates it at a given time interval.
+--
+-- @tparam command string The command.
```
It's `@tparam string command`, I think (but I'm not 100% sure).

actionless replied:
yup, travis confirmed that :)

psychon commented on lib/awful/widget/watch.lua:
```diff
@@ -0,0 +1,79 @@
+-- @return A textbox widget.
+function watch.new(command, timeout, callback, base_widget)
+    callback = callback or (function(widget, stdout, stderr) widget:set_text(stdout or stderr) end)
+    base_widget = base_widget or textbox()
```
Can these two lines be replaced with:
```
timeout = timeout or 5
base_widget = base_widget or textbox()
callback = callback or function(widget, stdout, stderr)
    widget:set_text(stdout) -- Your stdout or stderr does not work, the empty string is "trueish"
end
```

psychon commented on lib/awful/widget/watch.lua:
```diff
@@ -0,0 +1,79 @@
+-- @tparam timeout int (optional) The time interval at which the textbox will be updated. Default to 5.
+-- @tparam callback function (optional) The function that will be applied to the output before it is shown in the textbox.
+-- @param base_widget (optional) Defaults to `wibox.widget.textbox()`
+-- @return A textbox widget.
```
`@return The widget used by this watch`

actionless commented:
@psychon  btw, shouldn't exitcode be `int` instead of `number`?

psychon commented:
Yup, it should.

Elv13 commented on lib/awful/widget/watch.lua:
```diff
@@ -0,0 +1,84 @@
+---------------------------------------------------------------------------
+--- Watch widget.
```
There is complains on Reddit that we have no description for objects, it would be nice to have one (even maybe start to add screenshots with light themes (as the doc is black on white)

blueyed replied:
:D - do you have a link?

How should screenshots be linked / provided in the doc?

Elv13 replied:
See https://github.com/awesomeWM/awesome/pull/486

> :D - do you have a link?

There is more, but the most recent is:
https://www.reddit.com/r/awesomewm/comments/3yd4wg/is_there_an_alternative_api_documentation/

actionless replied:
i like the idea, to have an example code (fo default rc.lua) and screenshot of the widget attached

psychon replied:
Feel free to add that. Images go to `docs/images` and are linked to... relatively. `docs/03-declarative-layout.md` does `![Example4 screenshot](../images/widgetlayout2.png)`, I guess the same could work for this (as long as ldoc doesn't generate directories...)

actionless replied:
or should i reshot it with the light scheme?

psychon replied:
I don't mind this screenshot having another color theme. Feel free to leave it like this.

Elv13 commented on lib/awful/widget/watch.lua:
```diff
@@ -0,0 +1,84 @@
+
+local function easy_async_spawn(cmd, callback)
+  local stdout = ''
+  local stderr = ''
```
The indentation should be 4 space

actionless replied:
i will update this PR after/if this one will be merged: https://github.com/awesomeWM/awesome/pull/668

actionless commented:
updated

psychon commented on lib/awful/widget/watch.lua:
```diff
@@ -0,0 +1,62 @@
+--- Create a textbox that shows the output of a command
+-- and updates it at a given time interval.
+--
+-- @tparam string command The command.
```
`@tparam string|table cmd The command.`

psychon commented on lib/awful/widget/watch.lua:
```diff
@@ -0,0 +1,62 @@
+-- @tparam string command The command.
+--
+-- @tparam int timeout (optional) The time interval at which the textbox
+-- will be updated. Default to 5.
```
Default parameters work like this, I think: `@tparam[opt=5] int timeout The time interval....`

blueyed replied:
:+1:

psychon commented on lib/awful/widget/watch.lua:
```diff
@@ -0,0 +1,62 @@
+-- @tparam int timeout (optional) The time interval at which the textbox
+-- will be updated. Default to 5.
+--
+-- @tab callback Function (optional) The function that will be applied to
```
Could you try if `@tab[opt]` works? I'm not sure.

psychon replied:
Oh and this documents the callback as a table, not a function. @blueyed ideas? (Sorry, but you are the local ldoc guru (at least I think so))

blueyed replied:
`@tparam[opt] function callback The functiion that ...`?!

actionless replied:
but how then format callback's args? isn't `@tab` makes it happen?

psychon commented on lib/awful/widget/watch.lua:
```diff
@@ -0,0 +1,62 @@
+-- will be updated. Default to 5.
+--
+-- @tab callback Function (optional) The function that will be applied to
+-- the output before it is shown in the textbox.
```
I think this should be more explicit. Something like `the widget will display the return value of this function`?

actionless replied:
yup, in the original `watch` widget callback was just transforming (so docstring is living from that time), but i realized what it could be more useful to make it like now

psychon commented on lib/awful/widget/watch.lua:
```diff
@@ -0,0 +1,62 @@
+    timeout = timeout or 5
+    base_widget = base_widget or textbox()
+    callback = callback or function(widget, stdout, stderr, exitreason, exitcode) -- luacheck: no unused args
+        widget:set_text(stdout) -- Your stdout or stderr does not work, the empty string is "trueish"
```
Heh, sorry for the confusion. I didn't mean you to add that comment here, you can remove that. What I wanted to say was that your old version of this code doesn't work.

You did: `set_text(stdout or stderr)` (or something like that). However, `stdout` will never be `false` or `nil` and thus this always ignored `stderr` besides the code looking otherwise. The comment was meant to mention this to you, not something that I wanted you to copy into this PR.

Again, sorry for the confusion.

(Oh and thanks a lot for running this against Luacheck, that makes my attempts to automatically run Luacheck on Travis easier)

actionless replied:
oops :D blind copy-paste

psychon commented:
This looks fine to me and I only have minor comments (see above).

Does anyone have an idea on how a functional test for this could look like? My first idea would be a new test in `tests/` that just creates an instance of this widget like this:
```lua
local callbacks_done = 0
awful.widget.watch.new("echo hi", 0.1, function(widget, stdout, stderr, exitreason, exitcode)
  assert(widget == 42, widget)
  assert(stderr == "", stderr)
  assert(stdout == "hi", stdout)
  assert(exitreason == "exit", exitreason)
  assert(exitcode == 0, exitcode)
  callback_done = callbacks_done + 1
end, 42)
assert(callbacks_done == 1, callbacks_done)
require("_runner").run_steps{  function() return callbacks_done >= 2 end }
```
The above would check that the callback function really is invoked and that the timer (more or less) works.

(I only have this idea since `awful.mouse.finder` was broken for years and no one noticed.)

actionless commented:
test looks fine, i will try to integrate it

also i think it's worth asserting stdout in the test as well

actionless commented:
@blueyed so any suggestions regarding docstrings?

blueyed commented:
@actionless 
Commented above.

actionless commented:
i've done with the docs, will handle the tests later (hopefully this week)

actionless commented:
updated, funny what first i put it right (with `()`), but later on when i was debugging ldoc errors i removed it and did not put back

updated last commit and also added `@author` mentioning submitter of the original PR

actionless commented:
![2016-02-26--1456485986_727x281_scrot](https://cloud.githubusercontent.com/assets/1655669/13351026/2c2e55da-dc84-11e5-862f-6c005365290d.png)

`Mighty Ravendark` for the new release name :)

actionless commented:
@psychon added your test

psychon commented:
Looks good to me, but I guess this should be squashed into a single commit...?

Adding to next milestone anyway.

actionless commented:
squashed
